File: backend/src/main/java/com/andrew/hdss/models/Membership.java
package com.andrew.hdss.models;

import com.andrew.hdss.models.enums.MembershipEndType;
import com.andrew.hdss.models.enums.MembershipStartType;
import com.andrew.hdss.models.enums.RelationshipToHead;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDate;

@Getter
@Setter
@Entity
@Table(name = "memberships")
public class Membership extends SyncableEntity {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "individual_id", nullable = false)
    private Individual individual;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "household_id", nullable = false)
    private Household household;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RelationshipToHead relationshipToHead;

    @Column(nullable = false)
    private LocalDate startDate;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private MembershipStartType startType;

    // Null = this is the individual's current, open membership.
    private LocalDate endDate;

    @Enumerated(EnumType.STRING)
    private MembershipEndType endType;
}

File: android/app/src/main/java/com/andrew/hdss/data/models/Membership.kt
package com.andrew.hdss.data.models

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import com.andrew.hdss.data.models.enums.MembershipEndType
import com.andrew.hdss.data.models.enums.MembershipStartType
import com.andrew.hdss.data.models.enums.RelationshipToHead
import java.time.LocalDate

@Entity(
    tableName = "memberships",
    foreignKeys = [
        ForeignKey(
            entity = Individual::class,
            parentColumns = ["clientId"],
            childColumns = ["individualClientId"],
            onDelete = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = Household::class,
            parentColumns = ["clientId"],
            childColumns = ["householdClientId"],
            onDelete = ForeignKey.CASCADE
        )
    ],
    indices = [Index("individualClientId"), Index("householdClientId"), Index("serverId")]
)
data class Membership(
    @PrimaryKey
    val clientId: String,
    val serverId: Long?,
    val individualClientId: String,
    val householdClientId: String,
    val relationshipToHead: RelationshipToHead,
    val startDate: LocalDate,
    val startType: MembershipStartType,
    val endDate: LocalDate?,
    val endType: MembershipEndType?,
    val synced: Boolean = false
)